crates/bls-crypto/src/bls/public.rs

```diff
@@ -0,0 +1,210 @@
+        );
+        let y = x3b.sqrt().ok_or(io::Error::new(
+            io::ErrorKind::NotFound,
+            "couldn't find square root for x",
+        ))?;
+
+        let y_c0_big = y.c0.into_repr();
+        let y_c1_big = y.c1.into_repr();
+
+        let negy = -y;
+
+        let (bigger, smaller) = {
+            let half = Fq::modulus_minus_one_div_two();
+            if y_c1_big > half {
+                (y, negy)
+            } else if y_c1_big == half && y_c0_big > half {
+                (y, negy)
+            } else {
+                (negy, y)
+            }
+        };
+
+        let chosen_y = if y_over_half { bigger } else { smaller };
+        let pk = G2Affine::new(x, chosen_y, false);
+        Ok(PublicKey::from_pk(pk.into_projective()))
+    }
+
+    pub fn verify<H: HashToG1>(
+        &self,
+        message: &[u8],
+        extra_data: &[u8],
+        signature: &Signature,
+        hash_to_g1: &H,
+    ) -> Result<(), Box<dyn Error>> {
+        self.verify_sig(SIG_DOMAIN, message, extra_data, signature, hash_to_g1)
+    }
+
+    pub fn verify_pop<H: HashToG1>(
+        &self,
+        message: &[u8],
+        signature: &Signature,
+        hash_to_g1: &H,
+    ) -> Result<(), Box<dyn Error>> {
+        self.verify_sig(POP_DOMAIN, &message, &[], signature, hash_to_g1)
+    }
+
+    fn verify_sig<H: HashToG1>(
+        &self,
+        domain: &[u8],
+        message: &[u8],
+        extra_data: &[u8],
+        signature: &Signature,
+        hash_to_g1: &H,
+    ) -> Result<(), Box<dyn Error>> {
+        let pairing = Bls12_377::product_of_pairings(&vec![
+            (
+                signature.get_sig().into_affine().into(),
+                G2Affine::prime_subgroup_generator().neg().into(),
+            ),
+            (
+                hash_to_g1
+                    .hash::<Bls12_377Parameters>(domain, message, extra_data)?
+                    .into_affine()
+                    .into(),
+                self.pk.into_affine().into(),
+            ),
+        ]);
+        if pairing == Fq12::one() {
+            Ok(())
+        } else {
+            Err(BLSError::VerificationFailed)?
+        }
+    }
+}
+
+impl PartialEq for PublicKey {
+    fn eq(&self, other: &Self) -> bool {
+        // This byte-level equality operator differs from the (much slower) semantic
+        // equality operator in G2Projective.  We require byte-level equality here
+        // for HashSet to work correctly.  HashSet requires that item equality
+        // implies hash equality.
+        self.pk.x == other.pk.x && self.pk.y == other.pk.y && self.pk.z == other.pk.z
+    }
+}
+
+impl Hash for PublicKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // Only hash based on `y` for slight speed improvement
+        self.pk.y.hash(state);
+        // self.pk.x.hash(state);
+        // self.pk.z.hash(state);
+    }
+}
+
+impl ToBytes for PublicKey {
+    #[inline]
+    fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
```

Comment:
Ideally in the future we'd move here to `serialize_compressed`, but now it would be a breaking change